File: flutter_app/lib/models/wallet.dart
import 'package:equatable/equatable.dart';

/// Types of wallet transactions.
enum TransactionType { debit, credit, reward }

/// Represents the user's wallet.
class Wallet extends Equatable {
  final String id;
  final String userId;
  final double balance;
  final String currency;
  final bool isActive;

  const Wallet({
    required this.id,
    required this.userId,
    required this.balance,
    this.currency = 'USD',
    this.isActive = true,
  });

  factory Wallet.fromJson(Map<String, dynamic> json) {
    return Wallet(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      balance: (json['balance'] as num).toDouble(),
      currency: json['currency'] as String? ?? 'USD',
      isActive: json['is_active'] as bool? ?? true,
    );
  }

  @override
  List<Object?> get props => [id, userId, balance];
}

/// Represents a single wallet transaction.
class WalletTransaction extends Equatable {
  final String id;
  final TransactionType type;
  final double amount;
  final String description;
  final String? referenceId;
  final DateTime createdAt;

  const WalletTransaction({
    required this.id,
    required this.type,
    required this.amount,
    required this.description,
    this.referenceId,
    required this.createdAt,
  });

  factory WalletTransaction.fromJson(Map<String, dynamic> json) {
    return WalletTransaction(
      id: json['id'] as String,
      type: TransactionType.values.firstWhere(
        (t) => t.name == (json['type'] as String).toLowerCase(),
        orElse: () => TransactionType.debit,
      ),
      amount: (json['amount'] as num).toDouble(),
      description: json['description'] as String,
      referenceId: json['reference_id'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  @override
  List<Object?> get props => [id, type, amount, createdAt];
}

File: backend/src/routes/wallet.ts
import { Router, Response } from 'express';
import { authenticate, AuthRequest } from '../middleware/auth';

const router = Router();

/** GET /api/wallet */
router.get('/', authenticate, (req: AuthRequest, res: Response) => {
  res.json({
    wallet: { id: `wal_${req.user!.id}`, user_id: req.user!.id, balance: 1240.50, currency: 'USD', is_active: true },
    transactions: [
      { id: 'txn_1', type: 'debit', amount: 12.40, description: 'Ride Payment #ZR-882', reference_id: 'ride_882', created_at: '2024-03-24T14:45:00Z' },
      { id: 'txn_2', type: 'credit', amount: 500.00, description: 'Top Up from Bank', created_at: '2024-03-23T10:15:00Z' },
      { id: 'txn_3', type: 'reward', amount: 25.00, description: 'Weekly Bonus Reward', created_at: '2024-03-22T18:00:00Z' },
      { id: 'txn_4', type: 'debit', amount: 18.25, description: 'Ride Payment #ZR-712', reference_id: 'ride_712', created_at: '2024-03-21T11:30:00Z' },
    ],
  });
});

/** POST /api/wallet/top-up */
router.post('/top-up', authenticate, (req: AuthRequest, res: Response) => {
  const { amount } = req.body;
  res.json({ message: 'Top-up successful.', new_balance: 1240.50 + (amount || 0), transaction_id: `txn_${Date.now()}` });
});

/** POST /api/wallet/transfer */
router.post('/transfer', authenticate, (req: AuthRequest, res: Response) => {
  res.json({ message: 'Transfer successful.', transaction_id: `txn_${Date.now()}` });
});

/** POST /api/wallet/withdraw */
router.post('/withdraw', authenticate, (req: AuthRequest, res: Response) => {
  res.json({ message: 'Withdrawal initiated. Processing in 1-3 business days.', transaction_id: `txn_${Date.now()}` });
});

/** GET /api/wallet/transactions */
router.get('/transactions', authenticate, (req: AuthRequest, res: Response) => {
  res.json({ transactions: [], page: 1, total: 0 });
});

export default router;
